File: architecture/common/src/main/java/com/pacific/common/rx/MaybeUtil.java
package com.pacific.common.rx;

import io.reactivex.MaybeTransformer;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;
import java.util.concurrent.Executor;

public class MaybeUtil {

  private MaybeUtil() {
    throw new UnsupportedOperationException();
  }

  public static <T> MaybeTransformer<T, T> io() {
    return maybe -> maybe.subscribeOn(Schedulers.io())
        .observeOn(AndroidSchedulers.mainThread());
  }

  public static <T> MaybeTransformer<T, T> computation() {
    return maybe -> maybe.subscribeOn(Schedulers.computation())
        .observeOn(AndroidSchedulers.mainThread());
  }

  public static <T> MaybeTransformer<T, T> trampoline() {
    return maybe -> maybe.subscribeOn(Schedulers.trampoline())
        .observeOn(AndroidSchedulers.mainThread());
  }

  public static <T> MaybeTransformer<T, T> newThread() {
    return maybe -> maybe.subscribeOn(Schedulers.newThread())
        .observeOn(AndroidSchedulers.mainThread());
  }

  public static <T> MaybeTransformer<T, T> single() {
    return maybe -> maybe.subscribeOn(Schedulers.single())
        .observeOn(AndroidSchedulers.mainThread());
  }

  public static <T> MaybeTransformer<T, T> from(final Executor executor) {
    return maybe -> maybe.subscribeOn(Schedulers.from(executor))
        .observeOn(AndroidSchedulers.mainThread());
  }
}

File: architecture/arch-kotlin/src/main/java/com/pacific/arch/kotlin/rx/MaybeUtil.kt
package com.pacific.arch.kotlin.rx

import io.reactivex.MaybeTransformer
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import java.util.concurrent.Executor

class MaybeUtil private constructor() {
    companion object {
        
        @JvmStatic
        fun <T> io() = MaybeTransformer<T, T> {
            it.subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread())
        }

        @JvmStatic
        fun <T> computation() = MaybeTransformer<T, T> {
            it.subscribeOn(Schedulers.computation()).observeOn(AndroidSchedulers.mainThread())
        }

        @JvmStatic
        fun <T> trampoline() = MaybeTransformer<T, T> {
            it.subscribeOn(Schedulers.trampoline()).observeOn(AndroidSchedulers.mainThread())
        }

        @JvmStatic
        fun <T> newThread() = MaybeTransformer<T, T> {
            it.subscribeOn(Schedulers.newThread()).observeOn(AndroidSchedulers.mainThread())
        }

        @JvmStatic
        fun <T> single() = MaybeTransformer<T, T> {
            it.subscribeOn(Schedulers.single()).observeOn(AndroidSchedulers.mainThread())
        }

        @JvmStatic
        fun <T> from(executor: Executor) = MaybeTransformer<T, T> {
            it.subscribeOn(Schedulers.from(executor)).observeOn(AndroidSchedulers.mainThread())
        }
    }
}
